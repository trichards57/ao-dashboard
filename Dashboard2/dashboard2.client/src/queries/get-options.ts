import { queryOptions } from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export default function getOptions<T>(uri: string, queryKey: string[]) {
  return queryOptions({
    queryKey,
    queryFn: async () => {
      const response = await fetch(uri, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Network error");
      }

      return response.json() as Promise<T>;
    },
    staleTime: 10 * 60 * 1000,
    throwOnError: true,
  });
}
