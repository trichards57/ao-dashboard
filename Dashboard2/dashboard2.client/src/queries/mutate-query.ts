import { useMutation, useQueryClient } from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export default function useUpdate(uri: string, queryKey: string[]) {
  const queryClient = useQueryClient();

  return function useUpdateInner<TBody>(id: string) {
    return useMutation({
      mutationFn: async (body: TBody) => {
        const response = await fetch(`${uri}/${id}`, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(body),
        });

        if (!response.ok) {
          if (response.status === 404) {
            // eslint-disable-next-line @typescript-eslint/no-throw-literal
            throw notFound();
          }
          throw new Error("Failed to update.");
        }
      },
      onSuccess: () => {
        queryClient.invalidateQueries({ queryKey });
      },
    });
  };
}
