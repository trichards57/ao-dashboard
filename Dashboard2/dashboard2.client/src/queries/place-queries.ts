import {
  QueryClient,
  queryOptions,
  useSuspenseQuery,
} from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export type Region =
  | "All"
  | "EastOfEngland"
  | "EastMidlands"
  | "London"
  | "NorthEast"
  | "NorthWest"
  | "SouthEast"
  | "SouthWest"
  | "WestMidlands";

export function districtOptions(region: Region) {
  return queryOptions({
    queryKey: ["districts", region],
    queryFn: async () => {
      if (region == "All") {
        return [];
      }

      const response = await fetch(`/api/places/${region}`, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Failed to fetch districts.");
      }

      return response.json() as Promise<string[]>;
    },
    staleTime: 10 * 60 * 1000,
  });
}

export function hubOptions(region: Region, district: string) {
  return queryOptions({
    queryKey: ["hubs", region, district],
    queryFn: async () => {
      if (region == "All" || district.toUpperCase() == "ALL") {
        return [];
      }

      const response = await fetch(`/api/places/${region}/${district}`, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Failed to fetch hubs.");
      }

      return response.json() as Promise<string[]>;
    },
    staleTime: 10 * 60 * 1000,
  });
}

export function useDistricts(region: Region) {
  return useSuspenseQuery(districtOptions(region));
}

export function useHubs(region: Region, district: string) {
  return useSuspenseQuery(hubOptions(region, district));
}

export function preloadDistricts(queryClient: QueryClient, region: Region) {
  return queryClient.ensureQueryData(districtOptions(region));
}

export function preloadHubs(
  queryClient: QueryClient,
  region: Region,
  district: string,
) {
  return queryClient.ensureQueryData(hubOptions(region, district));
}
