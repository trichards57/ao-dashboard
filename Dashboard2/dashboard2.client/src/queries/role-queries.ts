import { QueryClient, useSuspenseQuery } from "@tanstack/react-query";

import getOptions from "./get-options";
import useUpdate from "./mutate-query";

export type Permission = "Deny" | "Read" | "Write";

export interface RolePermissions {
  id: string;
  name: string;
  permissions: Permission;
  vehicleConfiguration: Permission;
  vorData: Permission;
}

export interface RolePermissionsUpdate {
  permissions: Permission;
  vehicleConfiguration: Permission;
  vorData: Permission;
}

export function useUpdateRole(id: string) {
  return useUpdate("/api/roles", ["role"])<RolePermissionsUpdate>(id);
}

export function roleOptions(id: string) {
  return getOptions<RolePermissions>(`/api/roles/${id}`, ["role", id]);
}

export const allRoleOptions = getOptions<RolePermissions[]>("/api/roles", [
  "role",
]);

export function useRole(id: string) {
  return useSuspenseQuery(roleOptions(id));
}

export function useRoles() {
  return useSuspenseQuery(allRoleOptions);
}

export function preloadRole(queryClient: QueryClient, id: string) {
  return queryClient.ensureQueryData(roleOptions(id));
}

export function preloadRoles(queryClient: QueryClient) {
  return queryClient.ensureQueryData(allRoleOptions);
}
