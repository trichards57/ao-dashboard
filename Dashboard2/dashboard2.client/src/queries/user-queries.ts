import {
  useMutation,
  useQueryClient,
  UseSuspenseQueryOptions,
} from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export interface UserInfo {
  realName: string;
  userId: string;
  email: string;
  role: string;
  amrUsed: string;
  lastAuthenticated: string;
  otherClaims: Record<string, string>;
}

export interface UserWithRole {
  id: string;
  name: string;
  roleId: string;
  role: string;
}

export interface UserRoleUpdate {
  roleId: string;
}

export const useUpdateUser = (id: string) => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (user: UserRoleUpdate) => {
      const response = await fetch(`/api/users/${id}`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(user),
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Failed to update user.");
      }
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["user"] });
    },
  });
};

export const userSettings: (
  id: string,
) => UseSuspenseQueryOptions<UserWithRole> = (id: string) => ({
  queryKey: ["user", id],
  queryFn: async () => {
    const response = await fetch(`/api/users/${id}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      if (response.status === 404) {
        throw notFound();
      }
      throw new Error("Failed to fetch user settings.");
    }

    return response.json() as Promise<UserWithRole>;
  },
  staleTime: 10 * 60 * 1000
});

export const allUserOptions: UseSuspenseQueryOptions<UserWithRole[]> = {
  queryKey: ["user"],
  queryFn: async () => {
    const response = await fetch("/api/users", {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      throw new Error("Failed to fetch users.");
    }

    return response.json() as Promise<UserWithRole[]>;
  },
  staleTime: 10 * 60 * 1000
};

export const userMeOptions: UseSuspenseQueryOptions<UserInfo | null> = {
  queryKey: ["me"],
  queryFn: async () => {
    const response = await fetch("/api/users/me", {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      return Promise.resolve(null);
    }

    return response.json() as Promise<UserInfo>;
  },
  staleTime: 1000 * 60 * 5,
};
