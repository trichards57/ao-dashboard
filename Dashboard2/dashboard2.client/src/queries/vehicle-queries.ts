import { useMutation, useQueryClient } from "@tanstack/react-query";

export interface VehicleSettings {
  id: string;
  registration: string;
  callSign: string;
  hub: string;
  district: string;
  region: string;
  vehicleType: string;
  forDisposal: boolean;
}

export interface UpdateVehicleSettings {
  registration: string;
  hub: string;
  callSign: string;
  district: string;
  region: string;
  vehicleType: string;
  forDisposal: boolean;
}

export const useUpdateVehicle = () => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (vehicle: UpdateVehicleSettings) => {
      const response = await fetch("/api/vehicles", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(vehicle),
      });

      if (!response.ok) {
        throw new Error("Failed to update vehicle.");
      }
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["vehicle"] });
    },
  });
};

export const vehicleSettings = (id: string) => ({
  queryKey: ["vehicle", id],
  queryFn: async () => {
    const response = await fetch(`/api/vehicles/${id}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      throw new Error("Failed to fetch vehicle settings.");
    }

    return response.json() as Promise<VehicleSettings>;
  },
});

export const settingsOptions = (
  region: string | undefined,
  district: string | undefined,
  hub: string | undefined,
) => ({
  queryKey: ["settings", region ?? "All", district ?? "All", hub ?? "All"],
  queryFn: async () => {
    const response = await fetch(
      `/api/vehicles?region=${region ?? "All"}&district=${district ?? "All"}&hub=${hub ?? "All"}`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch vehicle settings.");
    }

    return response.json() as Promise<VehicleSettings[]>;
  },
});
