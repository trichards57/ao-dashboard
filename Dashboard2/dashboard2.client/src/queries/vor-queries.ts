export interface VorStatistics {
  totalVehicles: number;
  availableVehicles: number;
  vorVehicles: number;
  pastAvailability: Record<string, number>;
}

export interface VorStatus {
  id: string;
  region: string;
  hub: string;
  district: string;
  registration: string;
  callSign: string;
  summary: string;
  isVor: boolean;
  dueBack: string;
}

export const statisticsOptions = (
  region: string,
  district: string,
  hub: string,
) => ({
  queryKey: ["statistics", region ?? "All", district ?? "All", hub ?? "All"],
  queryFn: async () => {
    const response = await fetch(
      `/api/vor/statistics?region=${region ?? "All"}&district=${district ?? "All"}&hub=${hub ?? "All"}`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch hubs.");
    }

    return response.json() as Promise<VorStatistics>;
  },
});

export const statusOptions = (
  region: string,
  district: string,
  hub: string,
) => ({
  queryKey: ["status", region ?? "All", district ?? "All", hub ?? "All"],
  queryFn: async () => {
    const response = await fetch(
      `/api/vor?region=${region ?? "All"}&district=${district ?? "All"}&hub=${hub ?? "All"}`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch hubs.");
    }

    return response.json() as Promise<VorStatus[]>;
  },
});
