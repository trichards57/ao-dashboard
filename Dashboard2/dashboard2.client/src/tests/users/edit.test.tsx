import { useNavigate } from "@tanstack/react-router";
import { act, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ReactDOM from "react-dom/client";
import { Mock, beforeEach, describe, expect, it, vi } from "vitest";

import { RolePermissions, useRoles } from "../../queries/role-queries";
import {
  UserRoleUpdate,
  UserWithRole,
  useUpdateUser,
  useUser,
} from "../../queries/user-queries";
import { EditUser } from "../../routes/users/edit.$id";

vi.mock("../../queries/role-queries");
vi.mock("../../queries/user-queries");

const testRoles: RolePermissions[] = [
  {
    id: "a",
    name: "Administrator",
    permissions: "Write",
    vehicleConfiguration: "Deny",
    vorData: "Deny",
  },
  {
    id: "b",
    name: "Role 1",
    permissions: "Deny",
    vehicleConfiguration: "Read",
    vorData: "Read",
  },
  {
    id: "c",
    name: "Role 2",
    permissions: "Read",
    vehicleConfiguration: "Write",
    vorData: "Write",
  },
];

const testItem: UserWithRole = {
  id: "d",
  name: "User1",
  role: "Role 1",
  roleId: "b",
};

const testUpdate: UserRoleUpdate = {
  roleId: "c",
};

describe("Edit User Page", () => {
  beforeEach(() => {
    (useRoles as Mock).mockReturnValue({
      isLoading: false,
      data: testRoles,
    });
    (useUser as Mock).mockReturnValue({
      isLoading: false,
      data: testItem,
    });
    (useUpdateUser as Mock).mockReturnValue({
      mutateAsync: vi.fn(),
    });
  });

  it("renders without error", () => {
    const root = ReactDOM.createRoot(document.createElement("div"));
    expect(() => {
      act(() => {
        root.render(<EditUser id={testItem.id} isAdmin />);
        root.unmount();
      });
    }).not.toThrow();
  });

  it("renders correctly", () => {
    const { asFragment } = render(<EditUser id={testItem.id} />);
    expect(asFragment()).toMatchSnapshot();

    expect(screen.getByLabelText(/role/i)).toHaveValue(testItem.roleId);

    expect(useUser).toBeCalledWith(testItem.id);
  });

  it("renders admin correctly", () => {
    const { asFragment } = render(<EditUser id={testItem.id} isAdmin />);
    expect(asFragment()).toMatchSnapshot();

    expect(screen.getByLabelText(/role/i)).toHaveValue(testItem.roleId);
  });

  it("renders none correctly", () => {
    (useUser as Mock).mockReturnValue({
      isLoading: false,
      data: { ...testItem, roleId: "None" },
    });

    const { asFragment } = render(<EditUser id={testItem.id} isAdmin />);
    expect(asFragment()).toMatchSnapshot();

    expect(screen.getByLabelText(/role/i)).toHaveValue("None");
  });

  it("submits details to update api", async () => {
    const user = userEvent.setup();
    const mutateFn = vi.fn();
    const navigateFn = vi.fn();

    (useNavigate as Mock).mockReturnValue(navigateFn);
    (useUpdateUser as Mock).mockReturnValue({
      mutateAsync: mutateFn,
    });

    render(<EditUser id={testItem.id} />);

    await user.selectOptions(screen.getByLabelText(/role/i), testUpdate.roleId);

    await user.click(screen.getByRole("button", { name: /save/i }));

    expect(screen.getByLabelText(/role/i)).toBeValid();
    expect(mutateFn).toBeCalledWith(testUpdate);
    expect(navigateFn).toBeCalledWith({
      to: "/users",
    });
  });

  it("disabled items when running", async () => {
    const user = userEvent.setup();
    const mutateFn = vi.fn().mockReturnValue(new Promise(() => {}));
    const navigateFn = vi.fn();

    (useNavigate as Mock).mockReturnValue(navigateFn);
    (useUpdateUser as Mock).mockReturnValue({
      mutateAsync: mutateFn,
    });

    render(<EditUser id={testItem.id} />);

    await user.selectOptions(screen.getByLabelText(/role/i), testUpdate.roleId);

    await user.click(screen.getByRole("button", { name: /save/i }));

    expect(screen.getByLabelText(/role/i)).toBeDisabled();
    expect(screen.getByRole("button", { name: /save/i })).toBeDisabled();
  });
});
