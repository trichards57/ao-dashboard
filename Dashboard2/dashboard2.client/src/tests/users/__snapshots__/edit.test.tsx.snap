// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Edit User Page > renders admin correctly 1`] = `
<DocumentFragment>
  <h1
    class="title"
  >
    Update User1
  </h1>
  <form
    novalidate=""
  >
    <div
      class="field"
    >
      <label
        class="label"
        for="role"
      >
        Role
      </label>
      <div
        class="select is-fullwidth"
      >
        <select
          id="role"
          required=""
        >
          <option
            value="a"
          >
            Administrator
          </option>
          <option
            value="b"
          >
            Role 1
          </option>
          <option
            value="c"
          >
            Role 2
          </option>
        </select>
      </div>
    </div>
    <button
      class="button is-primary is-fullwidth"
      type="submit"
    >
      Save
    </button>
  </form>
</DocumentFragment>
`;

exports[`Edit User Page > renders correctly 1`] = `
<DocumentFragment>
  <h1
    class="title"
  >
    Update User1
  </h1>
  <form
    novalidate=""
  >
    <div
      class="field"
    >
      <label
        class="label"
        for="role"
      >
        Role
      </label>
      <div
        class="select is-fullwidth"
      >
        <select
          id="role"
          required=""
        >
          <option
            value="b"
          >
            Role 1
          </option>
          <option
            value="c"
          >
            Role 2
          </option>
        </select>
      </div>
    </div>
    <button
      class="button is-primary is-fullwidth"
      type="submit"
    >
      Save
    </button>
  </form>
</DocumentFragment>
`;

exports[`Edit User Page > renders none correctly 1`] = `
<DocumentFragment>
  <h1
    class="title"
  >
    Update User1
  </h1>
  <form
    novalidate=""
  >
    <div
      class="field"
    >
      <label
        class="label"
        for="role"
      >
        Role
      </label>
      <div
        class="select is-fullwidth"
      >
        <select
          id="role"
          required=""
        >
          <option
            value="None"
          >
            None
          </option>
          <option
            value="a"
          >
            Administrator
          </option>
          <option
            value="b"
          >
            Role 1
          </option>
          <option
            value="c"
          >
            Role 2
          </option>
        </select>
      </div>
    </div>
    <button
      class="button is-primary is-fullwidth"
      type="submit"
    >
      Save
    </button>
  </form>
</DocumentFragment>
`;
